src/decompression.jl

```
@@ -0,0 +1,59 @@
+module DecompressorStreams
+    using CodecZlib, Requires
+    export ArchiveFormat, decompressor_stream
+
+    abstract type ArchiveFormat{S} end
+
+    err_message(name) = string(
+        "Codec$name package is required to read $name-compressed RData files. ",
+        "Run Pkg.add(\"Codec$name\") to install it. Then in your code call ",
+        "using Codec$name before calling using RData."
+    )
+
+    function decompressor_stream(::Type{ArchiveFormat{S}}, io::IOStream) where S
+        if S == :bzip2
+            error(err_message("Bzip2"))
+        elseif S == :xz
+            error(err_message("Xz"))
+        else
+            throw(ArgumentError("Unsupported archive format: $S"))
+        end
+    end
+    decompressor_stream(::Type{ArchiveFormat{:gzip}}, io::IOStream) = GzipDecompressorStream(io)
+
+    function __init__()
+        @require CodecBzip2="523fee87-0ab8-5b00-afb7-3ecf72e48cfd" begin
+            using CodecBzip2
+            decompressor_stream(::Type{ArchiveFormat{:bzip2}}, io::IOStream) = Bzip2DecompressorStream(io)
+        end
+        @require CodecXz="ba30903b-d9e8-5048-a5ec-d1f5b0d4b47b" begin
+            using CodecXz
+            decompressor_stream(::Type{ArchiveFormat{:xz}}, io::IOStream) = XzDecompressorStream(io)
+        end
+    end
+end
+
+using .DecompressorStreams
```

I think moving the decompression code into separate file is a good idea.
But do we really need `DecompressionStreams` module? We export almost everything we define there.
For me it looks a little bit overengineering, plus it make this init logic harder to find (people expect it in `RData.jl` file)
Is there any technical reason not to do it in the main `__init__()`?

---

As I said, I tried out several designs and didn't like any of them very much. I don't like this one much either, for the same reasons as you. This was me trying to keep everything together. But it's not clear why the module exists, it's there so there is a local `__init__` method, then the `using .DecompressorStreams` triggers the `__init__`. It seemed at the time like it might be the "Julia" way of doing things.

> plus it make this init logic harder to find (people expect it in RData.jl file)

This is what I needed to know 👍 

I'll push a commit with a different approach.

I hope you don't mind a little _caring_ feedback from an experienced contributor: I suggest avoiding saying "overengineering" in reviews. I realise you didn't mean any harm by this and it is attributed to the code not me, but even as an experienced engineer and frequent contributor to open source, hearing it still makes me feel a bit embarrassed and discouraged 😞, then I need to shake off the feelings and pick myself back up. All I needed to know was the pros and cons of the design from a reader's perspective which you provided perfectly. ❤️

---

Thank you for the feedback! I'm very sorry if my wording has hurt your feelings.
I just wanted to clarify why I preferred one approach over the other. "Overengineering" seemed like a term that is neutral to the code quality (which is far better than most of the 1st PRs I've seen!). It was rather a reference to a noble aspiration of bringing more structure into this crazy world that motivates so many developers! ;)
I hope it would not detract you from further contributions to RData and/or other parts of Julia ecosystem.

---

Thank you and sorry for the delay, this deserves a proper reply, but too tired right now due to travels. Don't worry, it's all good.

---

Hi @alyst sorry for the delay in responding. Don't worry my feelings are ok. I wouldn't be put off contributing, and I think your response was ideal 👍

Unfortunately, I've heard people use the term "overengineering" as "[person] has poor judgement" or similar. Sometimes accompanied by "is a coding astronaut". It's why I briefly felt those emotions.

Reading carefully, I could see your intent was nothing like this. But I think you are right, in the context of this _text-only communication with strangers over the internet_, some contributors might misinterpret and silently go away.

I'm trying to get into the habit of assuming positive intent, and feeding back politely. If there's anything about my approach that could be improved let me know 😊